libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp

```diff
@@ -0,0 +1,469 @@
+        template <detail::async_rw_mutex_access_type AccessType>
+        struct sender;
+
+    public:
+        using read_type = void;
+        using readwrite_type = void;
+
+        using read_access_type =
+            detail::async_rw_mutex_access_wrapper<readwrite_type, read_type,
+                detail::async_rw_mutex_access_type::read>;
+        using readwrite_access_type =
+            detail::async_rw_mutex_access_wrapper<readwrite_type, read_type,
+                detail::async_rw_mutex_access_type::readwrite>;
+
+        using allocator_type = Allocator;
+
+        explicit async_rw_mutex(Allocator const& alloc = {})
+          : alloc(alloc){};
+        async_rw_mutex(async_rw_mutex&&) = default;
+        async_rw_mutex& operator=(async_rw_mutex&&) = default;
+        async_rw_mutex(async_rw_mutex const&) = delete;
+        async_rw_mutex& operator=(async_rw_mutex const&) = delete;
+
+        sender<detail::async_rw_mutex_access_type::read> read()
+        {
+            HPX_ASSERT(state);
+
+            if (prev_access == detail::async_rw_mutex_access_type::readwrite)
+            {
+                prev_state = state;
+                state =
+                    std::allocate_shared<detail::async_rw_mutex_shared_state,
+                        Allocator>(alloc);
+                prev_access = detail::async_rw_mutex_access_type::read;
+            }
+            return {prev_state, state};
+        }
+
+        sender<detail::async_rw_mutex_access_type::readwrite> readwrite()
+        {
+            HPX_ASSERT(state);
+
+            prev_state = state;
+            state = std::allocate_shared<detail::async_rw_mutex_shared_state,
+                Allocator>(alloc);
+            prev_access = detail::async_rw_mutex_access_type::readwrite;
+            return {std::move(prev_state), state};
+        }
+
+    private:
+        template <detail::async_rw_mutex_access_type AccessType>
+        struct sender
+        {
+            std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+            std::shared_ptr<detail::async_rw_mutex_shared_state> state;
+
+            template <template <typename...> class Tuple,
+                template <typename...> class Variant>
+            using value_types =
+                Variant<Tuple<detail::async_rw_mutex_access_wrapper<
+                    readwrite_type, read_type, AccessType>>>;
+
+            template <template <typename...> class Variant>
+            using error_types = Variant<std::exception_ptr>;
+
+            static constexpr bool sends_done = false;
+
+            template <typename R>
+            struct operation_state
+            {
+                std::decay_t<R> r;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> state;
+
+                operation_state(operation_state&&) = delete;
+                operation_state(operation_state const&) = delete;
```

Comment:
Should the assignment operators be deleted as well?

Reply:
Yep!